daemon/logger/adapter.go

```diff
@@ -127,9 +130,9 @@ func (a *pluginAdapterWithRead) ReadLogs(config ReadConfig) *LogWatcher {
 
 			select {
 			case watcher.Msg <- msg:
-			case <-watcher.WatchClose():
-				// make sure the message we consumed is sent
-				watcher.Msg <- msg
```

Comment:
Shouldn't we still send this in the case of producer gone?

Reply:
Well, this is a blocking send, and it will block here forever in case the consumer is gone.

Reply:
```go
select {
case <-watcher.Msg:
case <-watcher.WatchConsumerGone():
}
```

Reply:
Yes, this is what I was thinking as well, but in this case the code will look like this:

```go
select {
￼ 	case watcher.Msg <- msg:
	case <-watcher.WatchConsumerGone():
		return
	case <-watcher.WatchProducerGone():
		select {
		case watcher.Msg <- msg:
		case <-watcher.WatchConsumerGone():
		}
		return
}￼
```

Does it really have to look that ugly (in order to work correctly)?

Reply:
Could wrap this in a function... like `send(msg logger.Message, chMsg <-logger.Message, watcher logger.Watcher) <-chan struct`
With the return being a `chan struct{}` so you can keep it in a select if needed (not sure if it would be at the moment).

Reply:
From where I stand, it looks like if a consumer is gone, there is no sense in trying to send a message we have. So, the current code:

```go
			// send the message unless the consumer is gone
￼ 			select {
			case watcher.Msg <- msg:
			case <-watcher.WatchConsumerGone():
￼ 				return
￼ 			}
```

is sufficient and should not be made any more complicated than that.

Reply:
For ConsumerCone, yes. Originally this was checking producer, I think?
I will need to do a full review of the entire file at this point to make sure I'm understanding all the implications of these changes.

Reply:
Originally there was no distinction between the producer and the consumer. Both were calling `(LogWatcher *).Close()`, the first one was delivered, the second one was discarded. In case the first one was the Producer (meaning a container has stopped), the code we're discussing worked. In case the first one was the Consumer (meaning a log reader has exited), the code would block.

Now, with a distinction between the Producer and the Consumer, in this piece of code where we send a message to the Consumer, we are only interested if the Consumer is still there (to make sure we can send the message). In this place it does not make sense to check if the Producer is still there or not.